geo/geo.go

```diff
@@ -121,62 +155,44 @@ func (f *IPFilter) AllowLocation(loc *Location) bool {
 	return empty
 }
 
-func getLocationFromIP2C(ip string, retry int) (*Location, error) {
-	queryURL := "https://ip2c.nkn.org/" + ip
-	client := http.Client{
-		Timeout: 10 * time.Second,
+func getLocationFromProvider(ip string, p GeoProvider) Location {
+	loc, err := p.GetLocation(ip)
+	if err != nil {
+		log.Println(err)
 	}
+	return *loc
+}
 
-	i := 0
-	var resp *http.Response
-	var err error
-	for ; i < retry; i++ {
-		resp, err = client.Get(queryURL)
+func GetJsonFile(url, filename string) ([]byte, error) {
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		AWSFile, err := os.Create(filename)
 		if err != nil {
-			log.Println(err)
-			continue
+			return nil, err
 		}
-		break
-	}
-	if i == retry {
-		return &emptyLocation, err
-	}
+		defer AWSFile.Close()
 
-	defer resp.Body.Close()
-
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return &emptyLocation, err
+		resp, err := http.Get(url)
```

Comment:
Never use http.get, always creates a `http.Client` and set a timeout, otherwise this function may stuck forever. You can use a long timeout like 5 or 10 minutes.